openid-4-verifiable-presentations-1_0.md

```diff
@@ -616,11 +622,310 @@ The Wallet then validates the request as specified in OAuth 2.0 [@RFC6749].
+
+#### Selecting Credentials
+
+The following rules apply for selecting Credentials via `credentials` and `credential_sets`:
+
+- If `credential_sets` is not provided, the Verifier requests all
+  Credentials in `credentials` to be returned.
+- Otherwise, the Verifier requests all of the credential set queries in the `credential_sets` array
+  with `required` evaluated as true to be returned at a minimum and optionally any of the credential set queries 
+  with `required` evaluated as false.
+- For each credential set query inside the `credential_sets` array, in order to
+  satisfy the query, the Wallet MUST return a credential or credentials that
+  match to one of the `options` inside the object. It is typically
+  expected that the Wallet presents the End-User with a choice of which
+  credential to present if multiple credentials match the query.
+
+Credentials not matching the respective constraints expressed within
+`credentials` MUST NOT be returned, i.e., they are treated as if
+they would not exist in the Wallet.
+
+If the Wallet cannot deliver all non-optional Credentials requested by the
+Verifier according to these rules, it MUST NOT return any credential(s).
+
+## Format-specific Properties {#format_specific_properties}
+
+The following format-specific properties are defined:
+
+### IETF SD-JWT VC {#format_vc_sd_jwt}
+
+`vct_values`:
+: OPTIONAL. An array of strings that specifies allowed values for
+the type of the requested Verifiable Credential. All elements in the array MUST
+be valid type identifiers as defined in [@!I-D.ietf-oauth-sd-jwt-vc]. The Wallet
+MAY return credentials that inherit from any of the specified types, following
+the inheritance logic defined in [@!I-D.ietf-oauth-sd-jwt-vc].
+
+### Mobile Documents or mdocs (ISO/IEC 18013 and ISO/IEC 23220 series) {#format_mso_mdoc}
+
+`doctype_values`:
+: OPTIONAL. An array of strings that specifies allowed values for the
+doctype of the requested Verifiable Credential. All elements in the array MUST
+be valid doctype identifiers as defined in ISO 18013-5.
+
+
+## Claims Path Pointer {#claims_path_pointer}
+
+A claims path pointer is a pointer into the JSON structure of the Verifiable
+Credential, identifying one or more claims. A claims path pointer MUST be a
+non-empty array of strings and non-negative integers. A string value 
+indicates that the respective key is to be selected, the special string value `*` 
+indicates that all elements of the currently selected array(s) are to be selected;
+and a non-negative integer indicates that the respective index in an array is to be selected. The path
+is formed as follows:
+
+  - Start with an empty array.
+  - To address a particular claim within an object, append the key (claim name)
+    to the array.
+  - To address an element within an array, append the index to the array (as a
+    non-negative, 0-based integer).
+  - To address all elements within an array, append the string `*` to the array.
```

Comment:
this reads as a less-specified version of json path (which isn't really specified to begin with but [I see progress here](https://www.ietf.org/archive/id/draft-goessner-dispatch-jsonpath-00.html))

without this being testable (and tested) there's similar room for error/risk as were raised to using json path to begin with

Reply:
JsonPath is more complicated, requires complex string parsing, requires the implementation of many features that are not necessary to select a claim, and is a query language, not a pointer language (JsonPath would be appropriate to use). It is also somewhat inviting to severe security problems (the 'expression' syntax). That's why we didn't use it. We already received early feedback from implementers that they appreciate us not using JsonPath.

Reply:
if necessary, please open an issue to track/get implementer's feedback on this new approach compared to json path. but without merging this and getting people to try it out, we will not get those in the first place.

Reply:
https://github.com/openid/OpenID4VP/issues/286

I do not like the idea of merging controversial items and then trying to change them later. we all know that those later changes may not reach consensus and could be fought. this kind of tactic starts the ossification of controversial introductions to the spec which isn't so friendly.

Reply:
There has been extensive discussion on this. One of the main motivations for this was to get rid of JSON Path. We investigated JSON Pointer and identified things that could not be covered. Are you aware of other options? I think the proposed format is the simplest possible mechanism and was appreaciated by early developer's feedback